stl/src/atomic.cpp

```diff
@@ -4,33 +4,87 @@
+            __not_set,
+        };
+        static _STD atomic<_Shared_ptr_api_level> _Api_level{_Shared_ptr_api_level::__not_set};
+
+        _Shared_ptr_api_level _Current = _Api_level.load(_STD memory_order_acquire);
+        if (_Current == _Shared_ptr_api_level::__not_set) {
+            _Current          = _Shared_ptr_api_level::__has_nothing;
+            HMODULE _Kernel32 = GetModuleHandleW(L"Kernel32.dll");
+            if (_Kernel32 != nullptr) {
+                auto _Pfn_AcquireSRWLockExclusive = reinterpret_cast<decltype(&::AcquireSRWLockExclusive)>(
+                    GetProcAddress(_Kernel32, "AcquireSRWLockExclusive"));
+                auto _Pfn_ReleaseSRWLockExclusive = reinterpret_cast<decltype(&::ReleaseSRWLockExclusive)>(
+                    GetProcAddress(_Kernel32, "ReleaseSRWLockExclusive"));
+                if (_Pfn_AcquireSRWLockExclusive != nullptr && _Pfn_ReleaseSRWLockExclusive != nullptr) {
+                    _Table._Pfn_AcquireSRWLockExclusive.store(_Pfn_AcquireSRWLockExclusive, _STD memory_order_relaxed);
+                    _Table._Pfn_ReleaseSRWLockExclusive.store(_Pfn_ReleaseSRWLockExclusive, _STD memory_order_relaxed);
+                    _Current = _Shared_ptr_api_level::__has_srwlock;
+                }
+            }
+            _Api_level.store(_Current, _STD memory_order_release);
+        }
+        return reinterpret_cast<bool&>(_Current);
     }
-#else // _M_ARM
-    while (_interlockedbittestandset(&_Shared_ptr_flag, 0)) { // set bit 0
+#endif // ^^^ _STL_WIN32_WINNT < _STL_WIN32_WINNT_VISTA ^^^
+} // unnamed namespace
+
+_EXTERN_C
+
+_CRTIMP2_PURE void __cdecl _Lock_shared_ptr_spin_lock() { // spin until _Shared_ptr_flag successfully set
+#if _STL_WIN32_WINNT >= _STL_WIN32_WINNT_VISTA
+    AcquireSRWLockExclusive(_Shared_ptr_lock);
+#else // ^^^ _STL_WIN32_WINNT >= _STL_WIN32_WINNT_VISTA / _STL_WIN32_WINNT < _STL_WIN32_WINNT_VISTA vvv
+    if (_Acuire_srw_functions()) {
+        _Table._Pfn_AcquireSRWLockExclusive.load(_STD memory_order_relaxed)(&_Shared_ptr_lock);
+    } else {
+        while (_interlockedbittestandset(&_Shared_ptr_flag, 0) != 0) { // set bit 0
+            while (__iso_volatile_load32(reinterpret_cast<int*>(&_Shared_ptr_flag)) != 0) {
+                YieldProcessor();
+            }
+        }
     }
-#endif // _M_ARM
+#endif // ^^^ _STL_WIN32_WINNT < _STL_WIN32_WINNT_VISTA ^^^
 }
 
 _CRTIMP2_PURE void __cdecl _Unlock_shared_ptr_spin_lock() { // release previously obtained lock
-#ifdef _M_ARM
-    __dmb(_ARM_BARRIER_ISH);
-    __iso_volatile_store32(reinterpret_cast<volatile int*>(&_Shared_ptr_flag), 0);
-#else // _M_ARM
-    _interlockedbittestandreset(&_Shared_ptr_flag, 0); // reset bit 0
-#endif // _M_ARM
+#if _STL_WIN32_WINNT >= _STL_WIN32_WINNT_VISTA
+    ReleaseSRWLockExclusive(&_Shared_ptr_lock);
+#else // ^^^ _STL_WIN32_WINNT >= _STL_WIN32_WINNT_VISTA / _STL_WIN32_WINNT < _STL_WIN32_WINNT_VISTA vvv
+    if (_Acuire_srw_functions()) {

```

Comment:
If we keep this it shouldn't try to acquire again: if someone called this that means they already called lock

Reply:
I don't get it.

Reply:
What I mean is that this is trying to detect whether the SRW functions are present again, but we don't need to do that detection here. If a thread is calling `unlock`, it must have once already called `lock`, and when it called `lock` we did the detection attempt. So here we should be able to do something like:

```
const auto _Local_release = _Table._Pfn_ReleaseSRWLockExclusive.load(memory_order_relaxed);
if (_Local_release) { _Local_release(...); }
else { /* ... */ }
```

Reply:
Yes, makes sense